experimental/PropertyDDS/services/property-query/src/server/branch_write_queue.js

```diff
@@ -0,0 +1,757 @@
+            // Was it somewhere in the topology queue?
+            let correspondingInTopologyQueue = topologyWorkerQueue.find(({ task: queueItem }) =>
+              queueItem.guid === fr.guid
+            );
+
+            if (correspondingInTopologyQueue) {
+              // Queue already and bind the promise
+              let { task: foundTask, taskDp: foundDp } = correspondingInTopologyQueue;
+              foundTask.previouslyCompleted = true;
+              lastPromise = foundDp;
+              this._queueCommit(foundTask)
+                .then((res) => {
+                  logger.trace(`Done writing ${foundTask.guid}`);
+                  foundDp.resolve(res);
+                })
+                .catch((ex2) => {
+                  logger.trace(`Failed writing ${foundTask.guid}`, ex2);
+                  foundDp.reject(ex2);
+                });
+            } else {
+              // No?  Just add it as a pending promise in case someone later hooks to it
+              lastPromise = this._addPendingCommitPromise(task.branchGuid, fetchedResults.guid, this._queueCommit(fr));
+            }
+          });
+
+          minCommit = _.last(fetchedResults).guid;
+          // Await until the insertions are completed before fetching the next page as it can get big in memory.
+          await lastPromise;
+        } while (!allCommitsFound);
+      }
+
+      parentsFound.forEach((pf) => {
+        let { task: innerTask, taskDp: innerDp } = pf;
+        innerTask.previouslyCompleted = true;
+        this._queueCommit(innerTask)
+          .then((res) => {
+            logger.trace(`Done writing ${innerTask.guid}`);
+            innerDp.resolve(res);
+          })
+          .catch((ex2) => {
+            logger.trace(`Failed writing ${innerTask.guid}`);
+            innerDp.reject(ex2);
+          });
+      });
+    };
+
+    if (this._errorsPerBranch[task.branchGuid]) {
+      throw this._errorsPerBranch[task.branchGuid];
+    }
+
+    if (task.previouslyCompleted) {
+      logger.trace(`Item ${task.guid} was processed as part of compensation, skipping`);
+      return;
+    }
+
+    try {
+      logger.trace(`Validating uniqueness of ${task.guid}`);
+      await this._commitManager.getCommit(task.guid);
+      logger.trace(`Uniqueness of ${task.guid} failed.  Resolving with existing`);
+      // Commit already existing.  Resolve early
+      taskDp.resolve({ status: 'existing' });
+    } catch (ex) {
+
+      const change = {
+        guid: task.guid,
+        changeSet: new ChangeSet(task.changeSet),
+        referenceGuid: task.parentGuid,
+        remoteHeadGuid: task.meta &&  task.meta.remoteHeadGuid,
+        localBranchStart: task.meta && task.meta.localBranchStart
+      };
+
+      this._cache.set(change.guid, _.cloneDeep(change));
+
+      // Commit not already existing  This is where the fun begins
+      if (ex.statusCode === HTTPStatus.NOT_FOUND) {
+        let lastCommitGuid = await this._getLastCommitGuid(task.branchGuid);
```

Comment:
Are we sure that the commits will reach the query service in the right order and thus the last commit GUID here is always reliable?